### PR TITLE
Added TSQL support for CONVERT function with all its format possibilities

### PR DESCRIPTION
**Pull request by 10bas10**
Added support for the format options at the CONVERT function in T-SQL. Also corrected the datatype conversions for BLOBs and the default number of characters in case of no specified number of characters at the datatype.

### CODE REVIEW

File: sqlglot/dialects/tsql.py [hunk 1]
```diff
@@ -29,6 +96,10 @@ class Tokenizer(Tokenizer):
             "UNIQUEIDENTIFIER": TokenType.UNIQUEIDENTIFIER,
             "XML": TokenType.XML,
             "SQL_VARIANT": TokenType.VARIANT,
+            "NVARCHAR(MAX)": TokenType.TEXT,
+            "VARCHAR(MAX)": TokenType.TEXT,
+            "NVARCHAR": TokenType.VARCHAR,
+            "NCHAR": TokenType.CHAR,
```
Comment (GeorgeSittas) [inline sqlglot/dialects/tsql.py:102]: Why are these needed?

**Review by GeorgeSittas — COMMENTED** [review]
I don't think that the parser is the correct place to have the formatting logic. The parser should _just_ parse "CONVERT(NVARCHAR, x, 121)", and it can be formatted later if needed, eg. when generating sql code.

File: sqlglot/dialects/tsql.py (hunk 1 above)
Comment (10bas10) [reply sqlglot/dialects/tsql.py]: Thanks for the headsup. THe last two tokens are indeed not needed and i removed them now.

**Comment by 10bas10** [comment]
> I don't think that the parser is the correct place to have the formatting logic. The parser should _just_ parse "CONVERT(NVARCHAR, x, 121)", and it can be formatted later if needed, eg. when generating sql code.

Im not sure what you mean by this. From my perspective I am just parsing the CONVERT function. It includes a format however, so im translating that format to the Python format and then can be used at the generator for translation towards the corresponding dialect.

File: sqlglot/dialects/tsql.py [hunk 2]
```diff
@@ -9,6 +11,71 @@ class TSQL(Dialect):
     null_ordering = "nulls_are_small"
     time_format = "'yyyy-mm-dd hh:mm:ss'"
 
+    time_mapping = {
```
[the remark this answers is not in the document]
Comment (GeorgeSittas) [reply sqlglot/dialects/tsql.py:14]: I think hive is similar, you can try overriding that.

[the remark this answers is not in the document]
Comment (tobymao) [reply sqlglot/dialects/tsql.py:14]: let's have tsql keep it's own formatting and not reuse hive since they are unrelated

File: sqlglot/dialects/tsql.py [hunk 3]
```diff
@@ -37,11 +106,52 @@ class Parser(Parser):
...
         }
 
-        def _parse_convert(self):
+        def _parse_convert(self, strict):
             to = self._parse_types()
             self._match(TokenType.COMMA)
             this = self._parse_field()
-            return self.expression(exp.Cast, this=this, to=to)
+
+            var_length_datatype = [
```
Comment (tobymao) [inline sqlglot/dialects/tsql.py:114]: make this a constant outside of the function

File: sqlglot/dialects/tsql.py [hunk 4]
```diff
@@ -37,11 +106,52 @@ class Parser(Parser):
...
+                DataType.Type.NVARCHAR,
+                DataType.Type.VARCHAR,
+                DataType.Type.CHAR,
+                DataType.Type.NCHAR,
+            ]
+
+            # Retrieve length of datatype and override to default if not specified
+            if list_get(to.expressions, 0) is None and to.this in var_length_datatype:
+                to = exp.DataType.build(to.this, expressions=[exp.Literal(this="30", is_string=False)], nested=False)
```
Comment (tobymao) [inline sqlglot/dialects/tsql.py:123]: exp.Literal.number(30)

File: sqlglot/dialects/tsql.py [hunk 5]
```diff
@@ -37,11 +106,52 @@ class Parser(Parser):
...
+                    return self.expression(
+                        exp.StrToTime, this=this, format=exp.Literal(this=format_norm, is_string=True)
+                    )
+                # Check whether the convert entails a date to string format
+                elif to.this in var_length_datatype:
+                    return self.expression(
+                        exp.Cast if strict else exp.TryCast,
+                        to=to,
+                        this=self.expression(
+                            exp.TimeToStr, this=this, format=exp.Literal(this=format_norm, is_string=True)
```
Comment (tobymao) [inline sqlglot/dialects/tsql.py:145]: change all to Literal.string

File: sqlglot/parser.py [hunk 6]
```diff
@@ -2134,7 +2135,7 @@ def _parse_cast(self, strict):
 
         return self.expression(exp.Cast if strict else exp.TryCast, this=this, to=to)
 
-    def _parse_convert(self):
+    def _parse_convert(self, strict):
```
Comment (tobymao) [inline sqlglot/parser.py:2138]: strict is never used here, you should add the try convert logic in base

File: sqlglot/dialects/tsql.py (hunk 5 above)
Comment (10bas10) [reply sqlglot/dialects/tsql.py]: Changed to Literal.string and is included in the format_norm object now.

File: sqlglot/dialects/tsql.py (hunk 3 above)
Comment (10bas10) [reply sqlglot/dialects/tsql.py]: Placed it within the parser class now.

File: sqlglot/dialects/tsql.py (hunk 4 above)
Comment (10bas10) [reply sqlglot/dialects/tsql.py]: Changed it to a literal number.

File: sqlglot/parser.py (hunk 6 above)
Comment (10bas10) [reply sqlglot/parser.py:2138]: The function in the base class is now also using the strict parameter and will transpile to a TryCast or Cast.

File: sqlglot/dialects/tsql.py [hunk 7]
```diff
@@ -37,11 +106,44 @@ class Parser(Parser):
...
             this = self._parse_field()
-            return self.expression(exp.Cast, this=this, to=to)
+
+            # Retrieve length of datatype and override to default if not specified
+            if list_get(to.expressions, 0) is None and to.this in self.var_length_datatype:
+                to = exp.DataType.build(to.this, expressions=[exp.Literal.number(30)], nested=False)
+
+            # Check whether a conversion with format is applicable
+            if self._match(TokenType.COMMA):
+                format_norm = exp.Literal.string(TSQL.convert_format_mapping[int(self._parse_number().this)])
```
Comment (tobymao) [inline sqlglot/dialects/tsql.py:127]: int(self._parse_number().name)

Comment (tobymao) [reply sqlglot/dialects/tsql.py:127]: why convert to int and just make the format_mapping as tring?

File: sqlglot/dialects/tsql.py [hunk 8]
```diff
@@ -37,11 +106,44 @@ class Parser(Parser):
...
+
+                # Check whether the convert entails a string to date format
+                if to.this in [DataType.Type.DATE]:
```
Comment (tobymao) [inline sqlglot/dialects/tsql.py:130]: to.this == ...

File: sqlglot/dialects/tsql.py [hunk 9]
```diff
@@ -37,11 +106,44 @@ class Parser(Parser):
...
+                    return self.expression(exp.StrToDate, this=this, format=format_norm)
+                # Check whether the convert entails a string to datetime format
+                elif to.this in [DataType.Type.DATETIME]:
```
Comment (tobymao) [inline sqlglot/dialects/tsql.py:133]:
```suggestion
                elif to.this == DateType.Type.DATETIME
```

File: sqlglot/dialects/tsql.py [hunk 10]
```diff
@@ -37,11 +106,44 @@ class Parser(Parser):
             "CHARINDEX": exp.StrPosition.from_arg_list,
         }
 
-        def _parse_convert(self):
+        var_length_datatype = [
```
Comment (tobymao) [inline sqlglot/dialects/tsql.py:109]:
```suggestion
        VAR_LENGTH_DATATYPES = {
```

convert it into a set

**Comment by tobymao** [comment]
great pr @10bas10 thanks for doing this

**Comment by GeorgeSittas** [comment]
> Im not sure what you mean by this. From my perspective I am just parsing the CONVERT function. It includes a format however, so im translating that format to the Python format and then can be used at the generator for translation towards the corresponding dialect.

My understanding was incorrect earlier, sorry! What you're saying makes sense.

File: sqlglot/dialects/tsql.py (hunk 7 above)
Comment (10bas10) [reply sqlglot/dialects/tsql.py]: True, did just that now. Also built in a check on the validity of the users entry.

File: sqlglot/dialects/tsql.py (hunk 9 above)
Comment (10bas10) [reply sqlglot/dialects/tsql.py]: Fixed

File: sqlglot/dialects/tsql.py (hunk 10 above)
Comment (10bas10) [reply sqlglot/dialects/tsql.py]: Makes sense yes. On the capitalization of the objects i was not sure. Sometimes it is lower case, like in the dialect class. Changed it however to keep it in line within its own subclass.

**Comment by 10bas10** [comment]
Thanks Toby, great project and love working on it ;)

File: sqlglot/dialects/tsql.py (hunk 8 above)
Comment (10bas10) [reply sqlglot/dialects/tsql.py]: Also fixed

**Review by tobymao — APPROVED** [review]
